packages/snaps-ui/src/validation.ts

```diff
@@ -25,3 +30,70 @@ export function isComponent(value: unknown): value is Component {
 export function assertIsComponent(value: unknown): asserts value is Component {
   assertStruct(value, ComponentStruct, 'Invalid component');
 }
+
+const LINK_REGEX = /(?<protocol>[a-z]+:\/?\/?)(?<host>\S+?(?:\.[a-z]+)+)/giu;
+
+const ALLOWED_PROTOCOLS = ['https:', 'mailto:'];
+
+/**
+ * Searches for links in a sting and checks them against the phishing list.
+ *
+ * @param text - The text to verify.
+ * @param isOnPhishingList - The function that checks the link against the phishing list.
+ */
+export function assertLinksAreSafe(
+  text: string,
+  isOnPhishingList: (url: string) => boolean,
+) {
+  const links = text.match(LINK_REGEX);
+  if (links) {
+    links.forEach((link) => {
+      try {
+        const url = new URL(link);
+        assert(
+          ALLOWED_PROTOCOLS.includes(url.protocol),
+          `Protocol must be one of: ${ALLOWED_PROTOCOLS.join(', ')}.`,
+        );
+
+        const hostname =
+          url.protocol === 'mailto:'
+            ? url.pathname.split('@')[1]
+            : url.hostname;
+
+        assert(
+          !isOnPhishingList(hostname),
+          'The specified URL is not allowed.',
+        );
+      } catch (error) {
+        throw new Error(
+          `Invalid URL: ${
+            error instanceof AssertionError ? error.message : 'Invalid syntax.'
+          }`,
+        );
+      }
+    });
+  }
+}
+
+/**
+ * Searches for links in UI components and checks that the URL they are trying to
+ * pass in not in the phishing list.
+ *
+ * @param component - The custom UI component.
+ * @param isOnPhishingList - The function that checks the link against the phishing list.
+ */
+export function assertUILinksAreSafe(
+  component: Component,
+  isOnPhishingList: (url: string) => boolean,
+) {
+  const { type } = component;
+  if (type === NodeType.Panel) {
```

Comment:
@Mrtenz Is there an easier way we could do this with superstruct? If we could avoid manually traversing the components I would say that is ideal.

Reply:
I thought of creating a new refinement util for superstruct but it seems to be very painfull to always provide the function to the controller call everytime we need to validate the struct

Reply:
Yeah that would be ideal, but I don't think there is a good way to provide the validation function to the structs.